flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/physical/batch/BatchExecJoinBase.scala

```diff
@@ -70,4 +76,150 @@ abstract class BatchExecJoinBase(
+            return (false, null, null)
+          case _ => // do nothing
+        }
+      } else {
+        // cannot satisfy required hash distribution if requirement shuffle keys are not come from
+        // left side when Join is LOJ or are not come from right side when Join is ROJ.
+        return (false, null, null)
+      }
+    }
+    if (requiredLeftShuffleKeys.isEmpty) {
+      // the join can not satisfy the required hash distribution
+      // due to the required input shuffle keys are empty
+      return (false, null, null)
+    }
+
+    val (leftShuffleKeys, rightShuffleKeys) = if (joinType == JoinRelType.INNER &&
+      !requiredDistribution.requireStrict) {
+      (requiredLeftShuffleKeys.distinct, requiredRightShuffleKeys.distinct)
+    } else {
+      (requiredLeftShuffleKeys, requiredRightShuffleKeys)
+    }
+    (true,
+      FlinkRelDistribution.hash(ImmutableIntList.of(leftShuffleKeys: _*), requireStrict = true),
+      FlinkRelDistribution.hash(ImmutableIntList.of(rightShuffleKeys: _*), requireStrict = true))
+  }
+
+  /**
+    * Try to satisfy the given required traits on BroadcastJoin (including Broadcast-HashJoin and
+    * NestedLoopJoin).
+    *
+    * @param requiredTraitSet requirement traitSets
+    * @return Equivalent Join which satisfies required traitSet, return null if
+    *         requirement cannot be satisfied.
+    */
+  protected def satisfyTraitsOnBroadcastJoin(
+      requiredTraitSet: RelTraitSet,
+      leftIsBroadcast: Boolean): RelNode = {
+    val requiredDistribution = requiredTraitSet.getTrait(FlinkRelDistributionTraitDef.INSTANCE)
+    val keys = requiredDistribution.getKeys
+    val left = getLeft
+    val right = getRight
+    val leftFieldCnt = left.getRowType.getFieldCount
+    val canSatisfy = requiredDistribution.getType match {
+      case HASH_DISTRIBUTED | RANGE_DISTRIBUTED =>
+        // required distribution can be satisfied only if distribution keys all from
+        // non-broadcast side of BroadcastJoin
+        if (leftIsBroadcast) {
+          // all distribution keys must come from right child
+          keys.forall(_ >= leftFieldCnt)
+        } else {
+          // all distribution keys must come from left child
+          keys.forall(_ < leftFieldCnt)
+        }
+      // SINGLETON, BROADCAST_DISTRIBUTED, ANY, RANDOM_DISTRIBUTED, ROUND_ROBIN_DISTRIBUTED
+      // distribution cannot be pushed down.
+      case _ => false
+    }
+    if (!canSatisfy) {
+      return null
+    }
+
+    val keysInProbeSide = if (leftIsBroadcast) {
+      ImmutableIntList.of(keys.map(_ - leftFieldCnt): _ *)
+    } else {
+      keys
+    }
+
+    val inputRequiredDistribution = requiredDistribution.getType match {
+      case HASH_DISTRIBUTED =>
+        FlinkRelDistribution.hash(keysInProbeSide, requiredDistribution.requireStrict)
+      case RANGE_DISTRIBUTED =>
+        FlinkRelDistribution.range(keysInProbeSide)
+    }
+    // remove collation traits from input traits and provided traits
```

Comment:
why not building from an empty traitset and add the traits you can provide?

Reply:
we should keep the original provided traits, and we do not know how many kinds of trait in the traitset, although there are only 3 kinds of trait currently.

Reply:
I don't think so. You should check each trait you think you can provide very carefully. Simply keep the original provided traits looks very dangerous to me.